File: runme/mcgen/local/datan/datsrc/mindir.f
      SUBROUTINE MINDIR(X0,DIR,N,USERFN,FMIN,NSTEP,EPSILN)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(MAXSTP=1000,EPSDEF=1.D-8)
      PARAMETER(DELTAX=1.D-3,CUT=0.1D0,ZERO=0.D0)
      DIMENSION X0(N),DIR(N)
      EXTERNAL USERFN
C initialization
      NST=NSTEP
      IF(NST.LE.0) NST=MAXSTP
      EPS=EPSILN
      IF(EPS.LE.ZERO) EPS=EPSDEF
      XA=0.
      E=ABS(XA)
      IF(E.LT.CUT) E=CUT
      XB=E*DELTAX
      NSTIN=NST
C enclose minimum
      CALL MINENC(XA,XB,XC,YA,YB,YC,USERFN,NST,X0,DIR,N)
      IF(NST.LT.0) THEN
        NSTEP=-1
        GO TO 20
      END IF
      IF(XC.LT.XA)THEN
C reverse order of arguments
        DUM=XC
        XC=XA
        XA=DUM
      END IF
      NST=NSTIN
C locate minimum
      CALL MINCMB(XA,XC,EPS,USERFN,XMIN,FMIN,NST,X0,DIR,N)
      IF(NST.LT.0) THEN
        NSTEP=-1
        GO TO 20
      END IF
      DO 10 I=1,N
        X0(I)=X0(I)+XMIN*DIR(I)
   10 CONTINUE
   20 CONTINUE
      END

File: runme/mcgen/local/datan/datsrc/mindir.c
#include "datsrc.h"

void LIBFUNC mindir_(double P_T *x0, double const P_T *dir,
             integer const P_T *n,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), double P_T *fmin, integer P_T *nstep,
             double const P_T *epsiln)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    integer i, nstin, nst;
    double xmin, e, xa, xb, xc, ya, yb, yc, dum, eps;

/* initialization */
    /* Parameter adjustments */
    --dir;
    --x0;

    /* Function Body */
    nst = *nstep;
    if (nst <= 0) {
        nst = 1000;
    }
    eps = *epsiln;
    if (eps <= 0.) {
        eps = 1e-8;
    }
    xa = 0.f;
    e = abs(xa);
    if (e < .1) {
        e = .1;
    }
    xb = e * .001;
    nstin = nst;
/* enclose minimum */
    minenc_(&xa, &xb, &xc, &ya, &yb, &yc, userfn, &nst, &x0[1], &dir[1], n);
    if (nst < 0) {
        *nstep = -1;
        goto L20;
    }
    if (xc < xa) {
/* reverse order of arguments */
        dum = xc;
        xc = xa;
        xa = dum;
    }
    nst = nstin;
/* locate minimum */
    mincmb_(&xa, &xc, &eps, userfn, &xmin, fmin, &nst, &x0[1], &dir[1], n);
    if (nst < 0) {
        *nstep = -1;
        goto L20;
    }
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        x0[i] += xmin * dir[i];
    }
L20:
    return;
} /* mindir_ */

